### PR TITLE
Dependencies

### PR DESCRIPTION
**Pull request by OliverFriedrich**
Add dependency injection around existing code to make in usable in Laravel/Vue.js projects

### CODE REVIEW

**Comment by OliverFriedrich** [comment]
Would be great, if you could update the npm package. Than i just need to update my dependencies instead of replacing the package ...

File: src/pcre.js [hunk 1]
```diff
@@ -1,5 +1,15 @@
 // Declare global variable CodeMirror to avoid warnings in JSHint
-/* global CodeMirror */
+/* global CodeMirror, define */
+
+(function (mod) {
+  if (typeof exports === "object" && typeof module === "object") // CommonJS
```
Comment (xavierog) [inline src/pcre.js]: pcre.js is indented using tabs therefore I would appreciate if you could use tabs for these few extra lines.

File: src/pcre.js [hunk 2]
```diff
@@ -894,4 +894,4 @@ CodeMirror.defineMode('pcre', function(editor_options, mode_options) {
 CodeMirror.defineMIME('text/x-regex', 'pcre');
 CodeMirror.defineMIME('text/x-pcre-regex', 'pcre');
 
-});
\ No newline at end of file
+});
```
Comment (xavierog) [inline src/pcre.js:897]: Please squash this into the previous commit, it does not deserve a dedicated commit.

File: src/pcre.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Declare global variable CodeMirror to avoid warnings in JSHint
+// Declare global variables to avoid warnings in JSHint
```
Comment (xavierog) [inline src/pcre.js:1]: Not worth a dedicated commit either... reintegrate this into your second commit.

File: src/pcre.js [hunk 4]
```diff
@@ -1,5 +1,15 @@
 // Declare global variable CodeMirror to avoid warnings in JSHint
-/* global CodeMirror */
+/* global CodeMirror, define */
+
+(function (mod) {
+  if (typeof exports === "object" && typeof module === "object") // CommonJS
+    mod(require("codemirror/lib/codemirror"));
```
Comment (xavierog) [inline src/pcre.js]: Should it not be `../../lib/codemirror` instead of `codemirror/lib/codemirror` ? The idea here is to remain consistent with upstream CodeMirror modes, e.g. https://github.com/codemirror/codemirror5/blob/fc0440bc7563aead662492b0f1b0512637545442/mode/asn.1/asn.1.js#L6

**Review by xavierog — CHANGES_REQUESTED** [review]
Ok to integrate your changes. Especially, the CommonJS/AMD integration is something I completely ditched when writing this mode (and others) so your contribution on this matter is welcome.
However I have highlighted a series of details that must be fixed beforehand.

When you submit a pull request, the changes you made to the code will of course be reviewed and discussed. But the way you push these changes will equally be reviewed and discussed: number of commits + what each commit contains: is each change atomic? Are commit messages relevant and well written?
This makes sense considering Git enables you to completely rewrite history (`git rebase` is your friend) whenever needed. In the end, it is customary to submit an artificial history with perfect/elegant commits.

**Comment by xavierog** [comment]
> Would be great, if you could update the npm package. Than i just need to update my dependencies instead of replacing the package ...

Ok -- once your pull request is accepted, I intend to:

- "rebase & merge" your commits, 
- tag a new version
- update the npm package

File: src/pcre.js (hunk 1 above)
Comment (OliverFriedrich) [reply src/pcre.js]: no problem, have copied these lines from some existing mode script, so the spaces where already used there. But that can be changed

File: src/pcre.js (hunk 4 above)
Comment (OliverFriedrich) [reply src/pcre.js]: This would be true if this mode script would be part of the package "codemirror", but it isn't. So it is not possible/useful to use relative paths to access the elements in codemirror package.

So it is accessed by package name and path. CommonJS handles the loading ...

File: src/pcre.js (hunk 1 above)
Comment (OliverFriedrich) [reply src/pcre.js]: Indention of the file corrected. This leads to a complete replace of the original code as each line gets an additional tab in front of it...
